### DISCUSSION TITLE
Adding scaling with ranked subsampling (SRS) as an option for alpha diversity

### DISCUSSION THREAD

**Discussion by RachBioHaz** [cafferychen777/MicrobiomeStat#42] [Ideas]
Hi, just wondering if you would consider adding SRS as an option for calculating alpha diversity metrics. I love using your package, but using standard rarefication does not make sense for my data sets which have inherent differences in the biomass across the regions I need to compare.

**Comment by cafferychen777** [comment]
Dear @RachBioHaz,

Thank you for your valuable feedback and for using our package. I'm glad to hear that you find it useful for your research.

Regarding your suggestion to add SRS (Sequence Read Set) as an option for calculating alpha diversity metrics, I am very interested in exploring this possibility. However, I must admit that I am not entirely familiar with the specifics of this method, especially in the context of accounting for biomass differences across regions.

Could you kindly provide me with some additional information or references on this approach? Your insights would be immensely helpful in understanding how to implement this feature effectively.

I will make sure to include the development of this feature in our planning for the next period. Your input will be instrumental in ensuring that our package continues to meet the needs of our users.

Thank you once again for your suggestion and support. I look forward to hearing from you soon.

Best regards,
Chen YANG

**Comment by RachBioHaz** [comment]
Hi @cafferychen777 ,
not sure if we are referring to the same thing. The SRS I am referring to is Scaling with Ranked Subsampling. The paper for this approach can be fount here (https://peerj.com/articles/9593/) and the R documentation for thee package is here (https://cran.r-project.org/web/packages/SRS/SRS.pdf). 

Cheers
Rachele

**Reply by cafferychen777** [reply]
Dear @RachBioHaz,

Thank you for your clarification and for providing the references for Scaling with Ranked Subsampling (SRS). I appreciate your suggestion and the detailed information you've shared.

In response to your request, we've developed a new function that implements the SRS method for calculating alpha diversity indices. This function, called `mStat_calculate_alpha_diversity_SRS`, can be used as an alternative to our existing `mStat_calculate_alpha_diversity` function. Here's the basic usage of this new function:

```R
#' Calculate alpha diversity indices using Scaling with Ranked Subsampling (SRS)
#'
#' This function calculates alpha diversity indices (Shannon, Simpson, Observed Species) using the SRS method. 
#' The function takes an OTU table as input and returns a list containing the requested alpha diversity indices.
#'
#' @param x OTU table with taxa in rows and samples in columns.
#' @param alpha.name character vector containing the names of alpha diversity indices to calculate. 
#'        Possible values are: "shannon", "simpson", "observed_species".
#' @param Cmin numeric, the number of counts to which all samples will be normalized.
#' @return A list containing the requested alpha diversity indices.
#' @examples
#' \dontrun{
#' # Create example OTU table
#' otu.tab <- matrix(data = rpois(100, 5), nrow = 10, ncol = 10)
#' rownames(otu.tab) <- paste0("Taxon_", 1:10)
#' colnames(otu.tab) <- paste0("Sample_", 1:10)
#'
#' # Calculate alpha diversity indices using SRS
#' alpha.obj.srs <- mStat_calculate_alpha_diversity_SRS(x = otu.tab, 
#'                  alpha.name = c("shannon", "simpson", "observed_species"), 
#'                  Cmin = 100)
#' }
#' @export
mStat_calculate_alpha_diversity_SRS <- function(x, alpha.name, Cmin) {
  if (is.null(alpha.name)){
    return(NULL)
  }
  
  if (is.null(Cmin)) {
    stop("Cmin must be specified for the SRS method.")
  }
  
  # Perform SRS normalization
  x_srs <- SRS::SRS(data = x, Cmin = Cmin)
  x_transpose <- t(x_srs)
  
  alpha.obj <- lapply(alpha.name, function(index) {
    message(paste("Calculating", index, "diversity using SRS..."))
    result <- switch(index,
                     shannon = vegan::diversity(x_transpose, index = "shannon"),
                     simpson = vegan::diversity(x_transpose, index = "simpson"),
                     observed_species = vegan::specnumber(x_transpose),
                     {
                       warning(paste("Index", index, "is not supported with SRS method. Skipping."))
                       return(NULL)
                     }
    )
    if (!is.null(result)) {
      tibble::tibble(!!index := result, sample = rownames(x_transpose)) %>% 
        tibble::column_to_rownames("sample")
    } else {
      NULL
    }
  })
  
  # Remove NULL elements (unsupported indices)
  alpha.obj <- alpha.obj[!sapply(alpha.obj, is.null)]
  
  names(alpha.obj) <- alpha.name[alpha.name %in% names(alpha.obj)]
  
  message("SRS-based diversity calculations complete.")
  return(alpha.obj)
}
```

The parameters for this function are:

- `x`: OTU table with taxa in rows and samples in columns.
- `alpha.name`: A character vector containing the names of alpha diversity indices to calculate. Possible values are "shannon", "simpson", and "observed_species".
- `Cmin`: A numeric value specifying the number of counts to which all samples will be normalized in the SRS method.

The function returns a list containing the requested alpha diversity indices.

Here's an example of how to use it:

```R
# Create example OTU table
otu.tab <- matrix(data = rpois(100, 5), nrow = 10, ncol = 10)
rownames(otu.tab) <- paste0("Taxon_", 1:10)
colnames(otu.tab) <- paste0("Sample_", 1:10)

# Calculate alpha diversity indices using SRS
alpha.obj.srs <- mStat_calculate_alpha_diversity_SRS(
  x = otu.tab, 
  alpha.name = c("shannon", "simpson", "observed_species"), 
  Cmin = 100
)
```

This new function allows users to calculate alpha diversity using the SRS method, which is particularly useful for datasets with inherent differences in biomass across regions, as you mentioned.

We plan to include this new feature in our next version update. In the meantime, if you'd like to use this function, I'd be happy to provide you with the full function code.

Thank you again for your suggestion. It has helped us improve our package to better meet the needs of different users. If you have any questions or further suggestions, please don't hesitate to let me know.

Best regards,
Chen YANG
